### PR TITLE
CherryServers provider implementation

### PR DESCRIPTION
**Pull request by nacef-labidi**
Signed-off-by: Nacef LABIDI <nacef.labidi@gmail.com>

**What this PR does / why we need it**:
This PR implements a provider for [Cherry Servers](https://www.cherryservers.com/)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

```release-note
Cherry Servers provider implementation
```

### CODE REVIEW

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+	time.Sleep(30 * time.Second)
+
+	signer, err := ssh.ParsePrivateKey(sshPrivateKey)
+	sshClientConfig := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	conn, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", newSrv.IPAddresses[0].Address, "22"), sshClientConfig)
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: Ugh, not sure how I feel about this

Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: I did not have the choice without cloud-init support. Any ideas ?

Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Cherry servers have added cloud-init support to their API. No more tricky ssh stuff

File: Gopkg.toml [hunk 2]
```diff
@@ -144,6 +144,10 @@ required = [
   name = "github.com/docker/distribution"
   version = "v2.7.1"
 
+[[constraint]]
+  branch = "master"
```
Comment (xmudrii) [inline Gopkg.toml]: This should bind to a release or a commit to ensure that eventual breaking changes will not break the machine-controller

File: examples/cherryservers-machinedeployment.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                namespace: kube-system
+                name: machine-controller-cherryservers
+                key: token
+            plan: "Smart8"
+            location: "EU-East-1"
+            projectID: "<< YOUR_PROJECT_ID >>"
+            # Optional
+            labels:
+              my: label
+          operatingSystem: "Ubuntu 18.04 64bit"  # Or "CentOS 7 64bit"
```
Comment (xmudrii) [inline examples/cherryservers-machinedeployment.yaml]: I believe that this should be in the format of `ubuntu` and `centos` with machine-controller resolving to the correct image

File: pkg/cloudprovider/provider.go [hunk 4]
```diff
@@ -18,6 +18,7 @@ package cloudprovider
 
 import (
 	"errors"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/provider/cherryservers"
```
Comment (xmudrii) [inline pkg/cloudprovider/provider.go]: Move this import to the second group

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
+package cherryservers
+
+import (
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: Can we reorganize the imports to follow the order of:
* stdlib
* external dependencies
* machine-controller packages
* `k8s.io/*`
* `sigs.k8s.io/*`

Here is how it looks like for other packages

https://github.com/kubermatic/machine-controller/blob/9200b1809b024a90246d6496804409df6474f328/pkg/cloudprovider/provider/aws/provider.go#L19-L49

File: test/e2e/provisioning/all_e2e_test.go [hunk 6]
```diff
@@ -50,6 +50,7 @@ const (
 	OSUpgradeManifest      = "./testdata/machinedeployment-openstack-upgrade.yml"
 	invalidMachineManifest = "./testdata/machine-invalid.yaml"
 	kubevirtManifest       = "./testdata/machinedeployment-kubevirt.yaml"
+	cherryserversManifest = "./testdata/machinedeployment-cherryservers.yaml"
```
Comment (xmudrii) [inline test/e2e/provisioning/all_e2e_test.go]: This is not gofmt-ed

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	case providerconfig.OperatingSystemUbuntu:
+		return "Ubuntu 18.04 64bit", nil
+	case providerconfig.OperatingSystemCentOS:
+		return "CentOS 7 64bit", nil
+	}
+	return "", providerconfig.ErrOSNotSupported
+}
+
+func getClient(token string) *cherrygo.Client {
+	os.Setenv("CHERRY_AUTH_TOKEN", token)
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: In the documentation, you mention `CHERRYSERVERS_TOKEN`. Which one is supposed to be used?

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.DeleteMachineError,
+			Message: fmt.Sprintf("Could not delete machine, due to %v", err),
+		}
+	}
+
+	return false, nil
+}
+
+func NewKey() (privateKey []byte, publicKey string, err error) {
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: Is this function used anywhere? I wasn't able to find any reference, but I might be mistaken.

File: Gopkg.toml (hunk 2 above)
Comment (nacef-labidi) [reply Gopkg.toml]: Fixed

File: examples/cherryservers-machinedeployment.yaml (hunk 3 above)
Comment (nacef-labidi) [reply examples/cherryservers-machinedeployment.yaml]: Fixed

File: pkg/cloudprovider/provider.go (hunk 4 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 5 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider.go (hunk 4 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider.go]: Fixed

File: test/e2e/provisioning/all_e2e_test.go (hunk 6 above)
Comment (nacef-labidi) [reply test/e2e/provisioning/all_e2e_test.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 7 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: cherrygo client API needs CHERRY_AUTH_TOKEN to be set. I jave created CHERRYSERVERS_TOKEN to be inline with other providers. Then I make the provider set CHERRY_AUTH_TOKEN to the value of CHERRYSERVERS_TOKEN taken from manifest

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 8 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: My bad, not used anymore after removing ssh plumbing.

**Comment by nacef-labidi** [comment]
@xmudrii can you help me how to sign-off old commits that I forgot to sign, every time I try to amend an old commit I just create a new one while the old unsigned stays in the history

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+	client := getClient(c.Token)
+
+	servers, _, err := client.Servers.List(c.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, server := range servers {
+		if server.Hostname == machine.Name {
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: Is it safe to use the hostname as the mechanism for getting servers? Is it possible to happen that we have two servers with the same name? If yes, is this the case we need to be worried about?

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+				return nil, err
+			}
+			return &cherryServer{server: srv}, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, new types.UID) error {
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/cherryservers/provider.go]: Is this function not implemented intentionally? Do CherryServers support putting labels/tags on servers?

**Comment by xmudrii** [comment]
@nacef-labidi The easiest way to signoff all commits is to run rebase with the `--signoff` flag, such as:
```
git rebase --signoff origin/master
```

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 10 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: CherryServers does not support yet labels/tags

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 9 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Actually you can have 2 cherry servers with the same hostname. The fact that they don't support labels/tags make it tricky to fetch servers. What would you recommend ?

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.Token, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Token, "CS_TOKEN")
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed.

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int = 0
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go (hunk 7 above)
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+type Config struct {
+	TeamID    string
+	ProjectID string
+	Token     string
+	Plan      string
+	Location  string
+	Labels    map[string]string
+}
+
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+	}
+
+	c.TeamID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TeamID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"teamID\" field, error = %v", err)
+	}
+
+	c.Location, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Location)
+	if err != nil {
+		return nil, nil, err
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+func getClient(token string) *cherrygo.Client {
+	os.Setenv("CHERRY_AUTH_TOKEN", token)
+	client, err := cherrygo.NewClient()
+	if err != nil {
+		return nil
+	}
+	return client
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Fixed

File: pkg/cloudprovider/provider/cherryservers/provider.go [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+	}
+
+	var planID int = 0
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (nacef-labidi) [reply pkg/cloudprovider/provider/cherryservers/provider.go]: Validated beforehand with Validate method

**Comment by nacef-labidi** [comment]
Created a new cleaner PR #610 
Closing this one

**Pull request closed without merging**